pkg/rke2/psp/psp.go

```diff
@@ -0,0 +1,443 @@
+					return err
+				}
+			}
+		}
+		if _, err := cs.RbacV1().ClusterRoleBindings().Get(ctx, systemUnrestrictedRoleBindingName, metav1.GetOptions{}); err != nil {
+			if apierrors.IsNotFound(err) {
+				tmpl := fmt.Sprintf(systemUnrestrictedNodesRoleBindingTemplate, systemUnrestrictedRoleBindingName, systemUnrestrictedRoleName)
+				if err := deployClusterRoleBindingFromYaml(ctx, cs, tmpl); err != nil {
+					return err
+				}
+			}
+		}
+		if _, err := cs.RbacV1().ClusterRoleBindings().Get(ctx, systemUnrestrictedSvcAcctRoleBindingName, metav1.GetOptions{}); err != nil {
+			if apierrors.IsNotFound(err) {
+				tmpl := fmt.Sprintf(systemUnrestrictedServiceAcctRoleBindingTemplate, systemUnrestrictedSvcAcctRoleBindingName, systemUnrestrictedRoleName)
+				if err := deployRoleBindingFromYaml(ctx, cs, tmpl); err != nil {
+					return err
+				}
+			}
+		}
+		ns.SetAnnotations(map[string]string{namespaceAnnotationSystemUnrestricted: cisAnnotationValue})
+	}
+	return nil
+}
+
+// kubeConfigExists checks if the kubeconfig file
+// has been written.
+func kubeConfigExists(kc string) bool {
+	info, err := os.Stat(kc)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return !info.IsDir()
+}
+
+// SetPSPs sets the default PSP's based on the mode that RKE2 is running in. There is either CIS or non
+// CIS mode. For CIS mode, the globalRestricted and systemUnrestricted polices and associated roles and
+// bindings are created. For non CIS mode, globalRestricted, system-unrestricted, and global-unrestricted
+// policies and associated roles and bindings are loaded, if applicable.
+//
+// Each PSP has an annotation associated with it that we check to see exists before performing any
+// write operations. The annotation is created upon successful setting of PSPs.
+//
+// Load logic
+//
+// CIS:
+// - If the globalRestricted annotation does not exist, create PSP, role, binding.
+// - If the systemUnrestricted annotation does not exist, create the PSP, role, binding.
+// - If the globalUnrestricted annotation does not exist, check if PSP exists, and if so,
+//   delete it, the role, and the bindings.
+//
+// non-CIS:
+// - If the globalUnrestricted annotation does not exist, then create PSP, role, binding.
+// - If the systemUnrestricted annotation does not exist, then create PSP, role, binding.
+// - If the globalRestricted annotation does not exist, then check if the PSP exists and
+//   if it doesn't, create it. Check if the associated role and bindings exist and
+//   if they do, delete them.
+func SetPSPs(sCtx context.Context, ctx *cli.Context, k8sWrapTransport transport.WrapperFunc) error {
+	const (
+		kubeConfigPath = "/etc/rancher/rke2/rke2.yaml"
+		waitDelay      = time.Second * 1
+	)
+
+	// wait until the kubeconfig file is written
+	for !kubeConfigExists(kubeConfigPath) {
+		time.Sleep(waitDelay)
+	}
+
+	cs, err := newClient(kubeConfigPath, k8sWrapTransport)
+	if err != nil {
+		return err
+	}
+
+	for {
```

Comment:
Wrapping the entirety of this function in an infinite loop is concerning. What logic are you trying to retry? Is it just the waiting for the client or is it to do retries around specific client calls as well? 

In a non-error scenario, as best I can tell, you never exit this loop. Is that your intention?

Reply:
> In a non-error scenario, as best I can tell, you never exit this loop. Is that your intention?

That's not true...you do have a return nil in the last if block

Reply:
I've pushed error handling changes.